sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAlias.scala

```diff
@@ -0,0 +1,222 @@
+                  resolveByLateralAlias(u.nameParts, aliases.head.alias)
+                    .map(LateralColumnAliasReference(_, u.nameParts))
+                    .getOrElse(u)
+                case _ => u
+              }
+          }.asInstanceOf[NamedExpression]
+        }
+        val newProjectList = projectList.zipWithIndex.map {
+          case (a: Alias, idx) =>
+            val lcaWrapped = wrapLCAReference(a).asInstanceOf[Alias]
+            // Insert the LCA-resolved alias instead of the unresolved one into map. If it is
+            // resolved, it can be referenced as LCA by later expressions (chaining).
+            // Unresolved Alias is also added to the map to perform ambiguous name check, but only
+            // resolved alias can be LCA
+            aliasMap = insertIntoAliasMap(lcaWrapped, idx, aliasMap)
+            lcaWrapped
+          case (e, _) =>
+            wrapLCAReference(e)
+        }
+        p.copy(projectList = newProjectList)
+    }
+
+    // phase 2: unwrap
+    rewrittenPlan.resolveOperatorsUpWithPruning(
+      _.containsPattern(LATERAL_COLUMN_ALIAS_REFERENCE), UnknownRuleId) {
+      case p @ Project(projectList, child) if p.resolved
+        && projectList.exists(_.containsPattern(LATERAL_COLUMN_ALIAS_REFERENCE)) =>
+        // build the map again in case the project list changes and index goes off
+        // TODO one risk: is there any rule that strips off /add the Alias? that the LCA is resolved
+        //  in the beginning, but when it comes to push down, it really can't find the matching one?
+        //  Restore back to UnresolvedAttribute.
+        //  Also, when resolving from bottom up should I worry about cases like:
+        //  Project [b AS c, c + 1 AS d]
+        //  +- Project [1 AS a, a AS b]
+        //  b AS c is resolved, even b refers to an alias contains the lateral alias?
+        var aliasMap = CaseInsensitiveMap(Map[String, Seq[AliasEntry]]())
+        val referencedAliases = collection.mutable.Set.empty[AliasEntry]
+        def unwrapLCAReference(e: NamedExpression): NamedExpression = {
+          e.transformWithPruning(_.containsAnyPattern(LATERAL_COLUMN_ALIAS_REFERENCE)) {
+            case lcaRef: LateralColumnAliasReference if aliasMap.contains(lcaRef.nameParts.head) =>
+              val aliasEntry = aliasMap.get(lcaRef.nameParts.head).get.head
+              if (!aliasEntry.alias.containsPattern(LATERAL_COLUMN_ALIAS_REFERENCE)) {
+                // If there is no chaining, push down the alias and resolve the attribute by
+                // constructing a dummy plan
+                referencedAliases += aliasEntry
+                // Implementation notes (to-delete):
+                // this is a design decision whether to restore the UnresolvedAttribute, or
+                // directly resolve by constructing a plan and using resolveExpressionByPlanChildren
+                resolveByLateralAlias(lcaRef.nameParts, aliasEntry.alias).getOrElse(lcaRef)
```

Comment:
I'm confused, why not just return `lcaRef.ne`?

Reply:
I think it is safe to resolve it one more time, in case the alias or the field changes or is reconstructed (different exprId).